
# Machine Learning Project Proposal #

## Exploring Used Car Auction: *Prices and Predicting the Possible End Price*

### Simon Coessens, Rana Islek

### April 2024



### Necessary Libraries

In [ ]:
!pip install numpy==1.23.5
!pip install pandas==1.5.2
import warnings
warnings.filterwarnings('ignore')

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 53.1 MB/s eta 0:00:00
  Attempting uninstall: numpy
    Found existing installation: numpy 1.25.2
    Uninstalling numpy-1.25.2:
      Successfully uninstalled numpy-1.25.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
chex 0.1.86 requires numpy>=1.24.1, but you have numpy 1.23.5 which is incompatible.
pandas-stubs 2.0.3.230814 requires numpy>=1.25.0; python_version >= "3.9", but you have numpy 1.23.5 which is incompatible.


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 45.8 MB/s eta 0:00:00
  Attempting uninstall: pandas
    Found existing installation: pandas 2.0.3
    Uninstalling pandas-2.0.3:
      Successfully uninstalled pandas-2.0.3
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
cudf-cu12 24.4.1 requires pandas<2.2.2dev0,>=2.0, but you have pandas 1.5.2 which is incompatible.
google-colab 1.0.0 requires pandas==2.0.3, but you have pandas 1.5.2 which is incompatible.


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
from scipy.stats import skew
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
pd.set_option('display.max_columns', None)
%matplotlib inline
sns.set(rc={'figure.figsize':(20.7,8.27)})
sns.set_style("whitegrid")
sns.color_palette("dark")
plt.style.use("fivethirtyeight")

### Import Data

In [ ]:
dfcar = pd.read_csv('/content/drive/My Drive/ml-project/data/dfcar_processed.csv', on_bad_lines='skip')
dfcar.head(2)

vin  year  brand body_simple     model        trim made_in  \
0  wba3c1c51ek116351  2014    BMW       Sedan  3 Series  328I SULEV     DEU   
1  yv1612tb4f1310987  2015  VOLVO       Sedan       S60          T5     SWE   

  transmission  condition  color interior  car_age  odometer state  \
0    Automatic        4.5   Gray    Black        1    1331.0    CA   
1    Automatic        4.1  White    Black        0   14282.0    CA   

                                   seller  sellingprice    mmr  \
0  Financial Services Remarketing (Lease)         30000  31900   
1                 Volvo Na Rep/World Omni         27750  27500   

   is_sold_below_mmr day_of_week  is_weekend  hour_only month_year  \
0                  1         Thu           0          4     Jan-15   
1                  0         Thu           0          4     Jan-15   

    sale_date  sale_year   latitude   longitude  
0  2015-01-15       2015  37.271875 -119.270415  
1  2015-01-29       2015  37.271875 -119.270415

In [ ]:
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error,r2_score
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score


In [ ]:
columns_to_drop = ['year','vin','mmr', 'is_sold_below_mmr', 'day_of_week', 'is_weekend', 'is_weekend', 'month_year', 'sale_date', 'latitude', 'longitude', 'hour_only', 'seller']
dfcar = dfcar.drop(columns=columns_to_drop)
dfcar.head(2)

brand body_simple     model        trim made_in transmission  condition  \
0    BMW       Sedan  3 Series  328I SULEV     DEU    Automatic        4.5   
1  VOLVO       Sedan       S60          T5     SWE    Automatic        4.1   

   color interior  car_age  odometer state  sellingprice  sale_year  
0   Gray    Black        1    1331.0    CA         30000       2015  
1  White    Black        0   14282.0    CA         27750       2015

In [ ]:
categorical_features = ['brand', 'body_simple', 'model', 'trim', 'made_in', 'transmission', 'color', 'state', 'interior']
numerical_features = ['condition', 'odometer', 'car_age', 'car_age', 'sale_year']

preprocessor = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(drop='first', handle_unknown='ignore'), categorical_features)
    ])


In [ ]:
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsRegressor


X = dfcar.drop('sellingprice', axis=1)
y = dfcar['sellingprice']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

knn_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('knn', KNeighborsRegressor(n_neighbors=3))
])

knn_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', StandardScaler(),
                                                  ['condition', 'odometer',
                                                   'car_age', 'car_age',
                                                   'sale_year']),
                                                 ('cat',
                                                  OneHotEncoder(drop='first',
                                                                handle_unknown='ignore'),
                                                  ['brand', 'body_simple',
                                                   'model', 'trim', 'made_in',
                                                   'transmission', 'color',
                                                   'state', 'interior'])])),
                ('knn', KNeighborsRegressor(n_neighbors=3))])

In [ ]:
y_pred = knn_pipeline.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print(f'Mean Squared Error: {mse}')

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:202: UserWarning: Found unknown categories in columns [2, 3] during transform. These unknown categories will be encoded as all zeros
  warnings.warn(


Mean Squared Error: 4374680.876627351


In [ ]:
mae = mean_absolute_error(y_test, y_pred)
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
r2 = r2_score(y_test, y_pred)

print(f"Model Performance:")
print(f"Mean Absolute Error (MAE): {mae}")
print(f"Mean Squared Error (MSE): {mse}")
print(f"Root Mean Squared Error (RMSE): {rmse}")
print(f"R-squared (R2): {r2}")

Model Performance:
Mean Absolute Error (MAE): 1439.5740986462943
Mean Squared Error (MSE): 4374680.876627351
Root Mean Squared Error (RMSE): 2091.573779867053
R-squared (R2): 0.920086597975934


Model Performance:
Mean Absolute Error (MAE): 1439.5740986462943
Mean Squared Error (MSE): 4374680.876627351
Root Mean Squared Error (RMSE): 2091.573779867053
R-squared (R2): 0.920086597975934